src/CaptainHook.Common/ServiceBus/ServiceBusManager.cs

```
@@ -0,0 +1,117 @@
+using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.Management.ServiceBus.Fluent;
+using Microsoft.Azure.ServiceBus;
+using Microsoft.Azure.ServiceBus.Core;
+using Nito.AsyncEx;
+using Polly;
+using Polly.Retry;
+using ISubscription = Microsoft.Azure.Management.ServiceBus.Fluent.ISubscription;
+
+namespace CaptainHook.Common.ServiceBus
+{
+    /// <inheritdoc/>
+    public class ServiceBusManager : IServiceBusManager
+    {
+        private const int RetryCeilingSeconds = 30;
+
+        private readonly IMessageProviderFactory _factory;
+        private readonly AsyncLazy<IServiceBusNamespace> _serviceBusNamespace;
+        private readonly AsyncRetryPolicy<ITopic> _findTopicPolicy;
+
+        public ServiceBusManager(IMessageProviderFactory factory, AsyncLazy<IServiceBusNamespace> serviceBusNamespace)
+        {
+            _factory = factory;
+            _serviceBusNamespace = serviceBusNamespace;
+
+            static TimeSpan ExponentialBackoff(int x) => TimeSpan.FromSeconds(Math.Clamp(Math.Pow(2, x), 0, RetryCeilingSeconds));
+
+            _findTopicPolicy = Policy
+                .HandleResult<ITopic>(b => b == null)
+                .WaitAndRetryForeverAsync(ExponentialBackoff);
```

Do we want to keep retrying forever?

---

I'm not sure, the code was present there and I didn't question it. Seems like if we plan to try limited number of times, we would need to define the failure mechanism as well, like fail silently or throw the exception to the user?